From 8f1d8aded9ae4dc9d70ef82b1c02aea2a9bc07cb Mon Sep 17 00:00:00 2001
From: Jamin Lin <jamin_lin@aspeedtech.com>
Date: Tue, 29 Sep 2026 11:15:58 +0800
Subject: [PATCH v1 23/30] hw/arm/aspeed_ast1040: Add common init function for
 AST1040/AST1080 SoCs

AST1040 and AST1080 are the next-generation SoCs of AST1030 and AST1060,
respectively.

The AST1030 and AST1060 SoCs share common initialization through
aspeed_soc_ast10x0_init(). Similarly, introduce
aspeed_soc_ast10x0_g2_init() to provide common initialization for the
AST1040 and AST1080 SoCs.

Move the shared AST1040 initialization into the new common function so
that it can be reused when adding AST1080 support. The PECI and eSPI
controllers are only available on the AST1040, so their initialization
is kept in aspeed_soc_ast1040_init().

No functional changes.

Signed-off-by: Jamin Lin <jamin_lin@aspeedtech.com>
Change-Id: I19da2f22c9f78b777e033431558ab2030a2b3c14
---
 hw/arm/aspeed_ast1040.c | 55 ++++++++++++++++++++++++++++++++---------
 1 file changed, 44 insertions(+), 11 deletions(-)

diff --git a/hw/arm/aspeed_ast1040.c b/hw/arm/aspeed_ast1040.c
index 5c14d6197e..d7713782ac 100644
--- a/hw/arm/aspeed_ast1040.c
+++ b/hw/arm/aspeed_ast1040.c
@@ -87,9 +87,8 @@ static const int aspeed_soc_ast1040_irqmap[] = {
     [ASPEED_DEV_PECI]      = 164,
 };
 
-static const AspeedUnimpDevice aspeed_soc_ast1040_unimp_devs[] = {
+static const AspeedUnimpDevice aspeed_soc_ast10x0_g2_unimp_devs[] = {
     { "pwm",     "aspeed.pwm",   ASPEED_DEV_PWM,   0x10000 },
-    { "espi",    "aspeed.espi",  ASPEED_DEV_ESPI,  0x1000 },
     { "udc",     "aspeed.udc",   ASPEED_DEV_UDC,   0x4000 },
     { "jtag[0]", "aspeed.jtag0", ASPEED_DEV_JTAG0, 0x100 },
     { "jtag[1]", "aspeed.jtag1", ASPEED_DEV_JTAG1, 0x100 },
@@ -101,7 +100,13 @@ static const AspeedUnimpDevice aspeed_soc_ast1040_unimp_devs[] = {
     { "i3c", "aspeed.i3c", ASPEED_DEV_I3C, 0x8000 },
     { "caliptra", "aspeed.caliptra", ASPEED_DEV_CALIPTRA, 0x100000 },
 };
-QEMU_BUILD_BUG_ON(ARRAY_SIZE(aspeed_soc_ast1040_unimp_devs) > ASPEED_UNIMP_NUM);
+#define AST10X0_G2_UNIMP_NUM ARRAY_SIZE(aspeed_soc_ast10x0_g2_unimp_devs)
+
+static const AspeedUnimpDevice aspeed_soc_ast1040_unimp_devs[] = {
+    { "espi", "aspeed.espi", ASPEED_DEV_ESPI, 0x1000 },
+};
+#define AST1040_UNIMP_NUM ARRAY_SIZE(aspeed_soc_ast1040_unimp_devs)
+QEMU_BUILD_BUG_ON(AST10X0_G2_UNIMP_NUM + AST1040_UNIMP_NUM > ASPEED_UNIMP_NUM);
 
 static qemu_irq aspeed_soc_ast1040_get_irq(AspeedSoCState *s, int dev)
 {
@@ -111,11 +116,12 @@ static qemu_irq aspeed_soc_ast1040_get_irq(AspeedSoCState *s, int dev)
     return qdev_get_gpio_in(DEVICE(&a->armv7m), sc->irqmap[dev]);
 }
 
-static void aspeed_soc_ast1040_init(Object *obj)
+static void aspeed_soc_ast10x0_g2_init(Object *obj)
 {
     Aspeed10x0SoCState *a = ASPEED10X0_SOC(obj);
     AspeedSoCState *s = ASPEED_SOC(obj);
     AspeedSoCClass *sc = ASPEED_SOC_GET_CLASS(s);
+    const AspeedUnimpDevice *unimp;
     char typename[64];
     int i;
 
@@ -123,7 +129,7 @@ static void aspeed_soc_ast1040_init(Object *obj)
 
     s->sysclk = qdev_init_clock_in(DEVICE(s), "sysclk", NULL, NULL, 0);
 
-    /* AST1040 uses the AST2700 SCUIO model */
+    /* AST10x0 G2 uses the AST2700 SCUIO model */
     object_initialize_child(obj, "scu", &s->scu, TYPE_ASPEED_2700_SCUIO);
     qdev_prop_set_uint32(DEVICE(&s->scu), "silicon-rev", sc->silicon_rev);
 
@@ -135,7 +141,6 @@ static void aspeed_soc_ast1040_init(Object *obj)
     }
 
     object_initialize_child(obj, "adc", &s->adc, TYPE_ASPEED_2700_ADC);
-    object_initialize_child(obj, "peci", &s->peci, TYPE_ASPEED_PECI);
     object_initialize_child(obj, "gpio", &s->gpio, "aspeed.gpio-ast2700");
     for (i = 0; i < sc->sgpio_num; i++) {
         object_initialize_child(obj, "sgpio[*]", &s->sgpiom[i],
@@ -155,9 +160,28 @@ static void aspeed_soc_ast1040_init(Object *obj)
         object_initialize_child(obj, "spi[*]", &s->spi[i], typename);
     }
 
-    for (i = 0; i < ARRAY_SIZE(aspeed_soc_ast1040_unimp_devs); i++) {
-        object_initialize_child(obj, aspeed_soc_ast1040_unimp_devs[i].qom_name,
-                                &s->unimp[i], TYPE_UNIMPLEMENTED_DEVICE);
+    for (i = 0; i < AST10X0_G2_UNIMP_NUM; i++) {
+        unimp = &aspeed_soc_ast10x0_g2_unimp_devs[i];
+        object_initialize_child(obj, unimp->qom_name, &s->unimp[i],
+                                TYPE_UNIMPLEMENTED_DEVICE);
+    }
+}
+
+static void aspeed_soc_ast1040_init(Object *obj)
+{
+    AspeedSoCState *s = ASPEED_SOC(obj);
+    const AspeedUnimpDevice *unimp;
+    int i;
+
+    aspeed_soc_ast10x0_g2_init(obj);
+
+    object_initialize_child(obj, "peci", &s->peci, TYPE_ASPEED_PECI);
+
+    for (i = 0; i < AST1040_UNIMP_NUM; i++) {
+        unimp = &aspeed_soc_ast1040_unimp_devs[i];
+        object_initialize_child(obj, unimp->qom_name,
+                                &s->unimp[AST10X0_G2_UNIMP_NUM + i],
+                                TYPE_UNIMPLEMENTED_DEVICE);
     }
 }
 
@@ -325,13 +349,22 @@ static void aspeed_soc_ast1040_realize(DeviceState *dev_soc, Error **errp)
     }
 
     /* Unimplemented devices */
-    for (i = 0; i < ARRAY_SIZE(aspeed_soc_ast1040_unimp_devs); i++) {
-        unimp = &aspeed_soc_ast1040_unimp_devs[i];
+    for (i = 0; i < AST10X0_G2_UNIMP_NUM; i++) {
+        unimp = &aspeed_soc_ast10x0_g2_unimp_devs[i];
         aspeed_mmio_map_unimplemented(s->memory, SYS_BUS_DEVICE(&s->unimp[i]),
                                       unimp->region_name,
                                       sc->memmap[unimp->memmap_idx],
                                       unimp->size);
     }
+
+    for (i = 0; i < AST1040_UNIMP_NUM; i++) {
+        unimp = &aspeed_soc_ast1040_unimp_devs[i];
+        aspeed_mmio_map_unimplemented(s->memory,
+            SYS_BUS_DEVICE(&s->unimp[AST10X0_G2_UNIMP_NUM + i]),
+            unimp->region_name,
+            sc->memmap[unimp->memmap_idx],
+            unimp->size);
+    }
 }
 
 static void aspeed_soc_ast1040_class_init(ObjectClass *klass, const void *data)
-- 
2.53.0

